Feat / Player auto focus

royschut opened the pull request:
## Player auto focus (a11y)
- When the player opens, we should set its focus to the play/pause button from script, so that screen reader/keyboard users can logically keep navigating to interactive elements. I've done so from the `handleFirstFrame` callback, because from this point the buttons are available. I've used querySelectors, and I've made it more specific to the controlbar. [OTT-483]
- Moved the Cinema to a Modal, so that inert is handled. (With a nice opening Grow animation and a backdrop as a bonus.)
- Made sure the back button is always reachable for screen readers, by making the whole overlay disappear only using `opacity: 0`, instead of `display: none` or `visibility: hidden`. [OTT-830]
- Also Trailer back button remains visible using `opacity: 0`

[OTT-483]: https://videodock.atlassian.net/browse/OTT-483?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[OTT-830]: https://videodock.atlassian.net/browse/OTT-830?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

royschut committed:
feat(a11y): add auto focus to play/pause button of player

royschut committed:
refactor(a11y): remove obsolete overflow management

langemike commented:
Really nice, roy! But have you tested it on iOS and Android? On iOS it keeps the focus on the start watching button when the cinema is opened, and it reads the time progression. Really weird. I got it on video if you want to see it

On Android there is a layout issue. The top and bottom gets cut-off. See screenshot:

![SmartSelect_20240219_151221_Chrome](https://github.com/Videodock/ott-web-app/assets/1019129/55f7b45d-a244-4ff1-b49f-09d779129cb6)

I also have to suggestions regarding the back button:
1. Can we easily move the position in the DOM so the back button is before the controlbar? It feels a little bit weird to access to back-button after you have been through all the controlbar items. This makes the order illogical and I'm not sure if we will pass this on a audit.
2. Can we also add ESC-key support for closing the modal? This is also an requirement. I see you used `<Modal>` but the esc-key does not work.

royschut commented:
> Really nice, roy! But have you tested it on iOS and Android? On iOS it keeps the focus on the start watching button when the cinema is opened, and it reads the time progression. Really weird. I got it on video if you want to see it
> 
> On Android there is a layout issue. The top and bottom gets cut-off. See screenshot:

Great feedback Mike, thanks! 
- The iOS focus issue: that's strange, like inert is ignored. I will test and try to fix it on iOS. (should've done that indeed)
- Android overlap issue: I see it, I will fix it
- We could, but that means that the user has to go back (shift + tab) in order to reach the back button. I honestly doubt what is better. 🤔 @Videodock/jw-ott-front-end 
- Yes, I incorrectly assumed that would work immediately when using the Modal. I will fix it.

langemike commented:
> We could, but that means that the user has to go back (shift + tab) in order to reach the back button. I honestly doubt what is better. 🤔 @Videodock/jw-ott-front-end

Since it is expected that a document flows from top to bottom and left to right, it's a bit illogical you have to press SHIFT + TAB seeing this back-button in the top left corner. Of course it is really nice you were able to make it focusable at least :)

I hope we can easily fix this, if could potentially needs to be defended during an audit. Hopefully the the fix is easier ;) 

There are some guides written about this subject https://a11y-guidelines.orange.com/en/articles/test-wcag-132/ and  https://www.a11y-collective.com/glossary/tab-order/ It is also written on W3C, but it is a little bit vague about it. So maybe we can pass it.

royschut commented:
> I hope we can easily fix this, if could potentially needs to be defended during an audit. Hopefully the the fix is easier ;)

You have convinced me. I will timebox it.

ChristiaanScheermeijer commented on packages/ui-react/src/components/Player/Player.tsx:
```diff
@@ -72,7 +72,14 @@ const Player: React.FC<Props> = ({
   const handleUserActive = useEventCallback(onUserActive);
   const handleUserInactive = useEventCallback(onUserInActive);
-  const handleFirstFrame = useEventCallback(onFirstFrame);
+  const handleFirstFrame = useEventCallback(() => {
+    // a11y: auto focus to play/pause button
+    const playPauseButton = document.querySelector('.jw-controlbar [aria-label="Play"], .jw-button-container [aria-label="Pause"]') as HTMLDivElement | null;
+
+    playPauseButton?.focus();
+
+    onFirstFrame?.();
+  });
```
Does this also work (as expected) for the inline player variant?

royschut replied:
I hadn't checked (apologees!) but I did just now: works!

royschut committed:
refactor(player): fix player overlay appears under android controlbar

royschut committed:
feat(a11y): implement closing player with Escape key

royschut commented:
@langemike 
- iOS focus: That's an inconvenient bug, but it was there before this PR already. I'm not sure if it's something we can fix, we're correctly implementing inert. Create a known issue?
- Android height: fixed!
- Back button before controls: fixed!
- Close on Escape button: fixed!

I hope we're ready now. :-)

ChristiaanScheermeijer commented:
Is this still a draft PR @royschut?

royschut commented:
> Is this still a draft PR @royschut?

Not at all! Forgot. :D @ChristiaanScheermeijer

langemike commented:
Really nice fixes!

> * iOS focus: That's an inconvenient bug, but it was there before this PR already. I'm not sure if it's something we can fix, we're correctly implementing inert. Create a known issue?

I think we should discuss this with everyone involved. Maybe after the stand up of tomorrow? I think you right and this can be considered a "known issue", it's similar to https://videodock.atlassian.net/browse/DDSB-3438 which also goes wrong when a video player is on the page.

But the thing that worries me, is that the video player is a fundamental feature and in my test it's unusable. You cannot navigate to any video control element or even the close-button. Only when you force override the screen reader navigation. See video:

https://github.com/Videodock/ott-web-app/assets/1019129/275b1d8a-4bbb-4d11-8abc-88e2f4f42a61

I think we need to investigate this further or ask accessiblity.nl for consultancy next sprint. I think it's ok for now to merge. Since there is no easy fix or cause known.

Also on Android the screen reader behaves a little but wonky, but I can at least close the video with the back-button if i put some effort in to it. I would not consider this acceptable, but also this is out of scope of your work (player auto focus + close button).

Maybe these issues can only be fixed by JW player itself? Definitely requires some research. 

> Android height: fixed!
> * Back button before controls: fixed!
> * Close on Escape button: fixed!

All 3 verified! Thanks.

langemike reviewed (approved):
Only the snapshots needs to be updated!

And I approve. Though I think we need to discuss the a11y state of the video player with the team!

royschut commented:
> But the thing that worries me, is that the video player is a fundamental feature and in my test it's unusable.

I'm afraid I have to agree. I suggest I make a task in the bug sprint and pick it up end of this sprint if there is time left. (The iOS bug)

ChristiaanScheermeijer commented on packages/ui-react/src/components/Player/Player.tsx:
```diff
@@ -72,7 +72,14 @@ const Player: React.FC<Props> = ({
-  const handleFirstFrame = useEventCallback(onFirstFrame);
+  const handleFirstFrame = useEventCallback(() => {
+    // a11y: auto focus to play/pause button
+    const playPauseButton = document.querySelector('.jw-controlbar [aria-label="Play"], .jw-button-container [aria-label="Pause"]') as HTMLDivElement | null;
```
Opening the trailer while on the inline player page will target the wrong play/pause button (the first one found in the DOM).

Using the player container will ensure the correct player is queried:

```suggestion
    const playPauseButton = playerRef.current?.getContainer().querySelector('.jw-controlbar [aria-label="Play"], .jw-button-container [aria-label="Pause"]') as HTMLDivElement | null;
```

royschut replied:
Thanks. That's a side-case I haven't thought about.

royschut committed:
refactor(player): fix incorrect auto focus for trailer with inlineplayer

ChristiaanScheermeijer reviewed (approved):
Nice @royschut!

MelissaDTH reviewed (approved):
LGTM!

royschut committed:
refactor(tests): update cinema snapshot
